### ISSUE TITLE
negative drifts when cycling using MAD-X 5.02.04

### ISSUE THREAD

**Issue by coldfix** [MethodicalAcceleratorDesign/MAD-X#317]
*Issue migrated from trac ticket # 317*

**component:** core_sequence | **priority:** major | **resolution:** invalid | **keywords:** SEQEDIT FLATTEN CYCLE circular
___
#### 2015-01-07 09:11:42: rdemaria commented

Cycling without flatten generates broken sequence (it works with pro release). See example...
 

___
#### 2015-01-07 09:16:54: ghislain commented

Hello Riccardo, 

I do not have access to /home/rdemaria/work/lhc_2015/opt_med db5
Is the V6.5.seq file a standard one ? if not can you attach it to the ticket with all necessary CALLed files. 

Cheers.

___
#### 2015-01-07 09:16:54: ghislain changed owner from ** to *ghislain*

___
#### 2015-01-07 09:16:54: ghislain changed status from *new* to *assigned*

___
#### 2015-01-07 09:17:01: ghislain changed status from *assigned* to *accepted*

___
#### 2015-01-07 09:23:19: rdemaria uploaded file [`job_bugcycle.madx`](https://raw.githubusercontent.com/MethodicalAcceleratorDesign/madx-tickets/master/attachments/317/job_bugcycle.madx) (0.6 KiB)

___
#### 2015-01-07 09:25:05: rdemaria commented

It is the standard one. I also changed the path in the example

___
#### 2015-01-07 11:29:17: ghislain commented

The observed behavior showing the negative drift is correct given the way that the LHC sequence is specified.
In order to avoid this negative drift one must do a FLATTEN before the CYCLE.

The reason why this used to work is that there was an assumption in routine all_node_pos in mad_seq,c that machines are circular with the addition of sequence length if the resulting position of an element is negative:
    if ((node->position = get_node_pos(node, sequ)) < zero)   node->position += sequence_length(sequ);

This is clever but only works for CIRCULAR machines. If the user makes input mistake for an open line, eg transfer line, the node position will be actually wrongly calculated and elements positioned at the other end of the line, instead of reporting a negative drift. There are also other side effects of this when inserting a subsequence within a sequence with a REFPOS given.

The modification to remove this assumption was introduced on 2014-Mar-26 in release [r4728](../commit/c0db690bf43edbf081e2e31e56fa27eec1d355ab) and has been in dev release since 5.02.01. Hence it is correct that 5.02.00 still has the old behavior. 

I will make sure that the documentation for SEQEDIT includes a recommendation of doing a FLATTEN before doing a CYCLE.

In summary this is not a bug but the observed behavior results from a bug fix of some 9 months ago.

___
#### 2015-01-07 11:29:17: ghislain set resolution to *invalid*

___
#### 2015-01-07 11:29:17: ghislain changed status from *accepted* to *closed*

___
#### 2015-01-08 10:13:04: @ldeniau commented

date: 2015.01.08, from ldeniau

One cheap alternative in the presence of many CYCLE command (also suggested by Ghislain) is to flatten the sequence right after loading it. Hence all subsequent editions including cycling will work properly.
